### DISCUSSION TITLE
Evaluate Profile-Guided Optimization (PGO) and LLVM BOLT

### DISCUSSION THREAD

**Discussion by zamazan4ik** [bottlerocket-os/bottlerocket#3454] [Ideas]
Hi!

Recently I did many Profile-Guided Optimization (PGO) benchmarks on multiple projects - the results are available [here](https://github.com/zamazan4ik/awesome-pgo). Here you can find different applications from different domains that were accelerated with PGO: operating systems (like Linux and Windows kernels), virtual machines (like QEMU and CrosVM), compilers, gRPC workloads, benchmark tools, databases, and much more. So that's why I think it's worth trying to apply PGO to Bottlerocket.

I can suggest the following things to do:
* Evaluate PGO's applicability to Bottlerocket.
* If PGO helps to achieve better performance - add a note to Bottlerocket's documentation about that. In this case, users and maintainers will be aware of another optimization opportunity for Bottlerocket.
* Provide PGO integration into the build scripts. It can help users and maintainers easily apply PGO for their own workloads.
* Optimize prebuilt Bottlerocket binaries with PGO.

After PGO, I can suggest evaluating [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md) as an additional optimization step after PGO.

For the Rust projects, I recommend starting with [cargo-pgo](https://github.com/Kobzol/cargo-pgo).

**Comment by stmcginnis** [comment]
Thanks for raising this @zamazan4ik. PGO looks interesting - I'll have to look in to it and see how it works with Bottlerocket. If it is useful, that looks like a good plan!
